flink-staging/flink-ml/src/test/scala/org/apache/flink/ml/feature/extraction/FeatureHasherSuite.scala

```diff
@@ -0,0 +1,245 @@
+
+class FeatureHasherSuite
+  extends FlatSpec
+  with Matchers
+  with FlinkTestBase {
+
+  behavior of "Flink's Feature Hasher"
+
+  import FeatureHasherData._
+
+  it should "transform a sequence of strings into a sparse feature vector of given size" in {
+    val env = ExecutionEnvironment.getExecutionEnvironment
+
+    env.setParallelism(2)
+
+    for (numFeatures <- numFeaturesTest) {
+      val inputDS = env.fromCollection(input)
+
+      val transformer = FeatureHasher()
+        .setNumFeatures(numFeatures)
+
+      val transformedDS = transformer.transform(inputDS)
+      val results = transformedDS.collect()
+
+      for ((result, expectedResult) <- results zip expectedResults(numFeatures)) {
+        result.equalsVector(expectedResult) should be(true)
+      }
+    }
+  }
+
+  it should "transform a sequence of strings into a sparse feature vector of given size," +
+    "with non negative entries" in {
+    val env = ExecutionEnvironment.getExecutionEnvironment
+
+    env.setParallelism(2)
+
+    for (numFeatures <- numFeaturesTest) {
+      val inputDS = env.fromCollection(input)
+
+      val transformer = FeatureHasher()
+        .setNumFeatures(numFeatures).setNonNegative(true)
+
+      val transformedDS = transformer.transform(inputDS)
+      val results = transformedDS.collect()
+
+      for ((result, expectedResult) <- results zip expectedResultsNonNegative(numFeatures)) {
+        result.equalsVector(expectedResult) should be(true)
+      }
+    }
+  }
+
+  it should "transform a sequence of strings into a sparse feature vector of default size," +
+    " when parameter is less than 1" in {
+    val env = ExecutionEnvironment.getExecutionEnvironment
+
+    env.setParallelism(2)
+
+    val inputDS = env.fromCollection(input)
+
+    val numFeatures = 0
+
+    val transformer = FeatureHasher()
+      .setNumFeatures(numFeatures).setNonNegative(false)
+
+    val transformedDS = transformer.transform(inputDS)
+    val results = transformedDS.collect()
+
+    for (result <- results) {
+      result.size should equal(Math.pow(2, 20).toInt)
+    }
+  }
+}
+
+object FeatureHasherData {
+
+  val input = Seq(
+    "Two households both alike in dignity".split(" ").toSeq,
+    "In fair Verona where we lay our scene".split(" ").toSeq,
+    "From ancient grudge break to new mutiny".split(" ").toSeq,
+    "Where civil blood makes civil hands unclean".split(" ").toSeq,
+    "From forth the fatal loins of these two foes".split(" ").toSeq
+  )
+
+  /* 2^30 features can't be tested right now because the implementation of Vector.equalsVector
+  performs an index wise comparison on the two vectors, which takes approx. forever */
+  val numFeaturesTest = Seq(Math.pow(2, 4).toInt, Math.pow(2, 5).toInt, 1234,
+    Math.pow(2, 16).toInt, Math.pow(2, 20).toInt) //, Math.pow(2, 30).toInt)
+
+  val expectedResults = List(
+    16 -> List(
+      SparseVector.fromCOO(16, Map((0, 1.0), (1, 1.0), (2, -1.0), (14, -1.0))),
```

Comment:
you can use arrow notation for the `pair` entries in the `Map` constructor, e.g.,

```scala
Map(0 -> 1.0, 1 -> 1.0, ...)
```